connaisseur/container.py

```diff
@@ -0,0 +1,15 @@
+from connaisseur.image import Image
+
+
+class Container:
+    image: Image
+    index: int
+    container_type: str
+
+    def __init__(self, image: str, index: int, container_type: str):
+        self.image = Image(image)
+        self.index = index
+        self.container_type = container_type
+
+    def __eq__(self, other):
```

Comment:
Is there a reason for this equality? Given that this is explicitily encapsulates the index in addition to the Image, it seems weird that it would define equality the same way. If there's a reason, I'd add a comment

Reply:
https://github.com/sse-secure-systems/connaisseur/pull/157#discussion_r627288860

twas your idea to overwrite the equality function. this also makes the code for checking whether the current container already has been validated previously really **sexy**.

```python
 if container in admission_request.wl_object.parent_containers:
     msg = f'automatic child approval for "{str(container.image)}".'
     logging.info(__create_logging_msg(msg, **logging_context))
     continue
```

Reply:
Then I'd still add a comment to the equality telling that equality is also used for containers to check whether their image is the same and there would be no other apparent reason to compare containers with the intention of having the index play a part
As I said above, I'm not opposed, but it looks weird since its a data structure that explicitly has an additional parameter that is then not used. Hence, I would say it warrants a comment

Reply:
Thinking about it, I'd say the correct way of doing this would be to keep the index outside of the `Container` object. `Container` objects are members of [lists like this](https://github.com/sse-secure-systems/connaisseur/pull/157/commits/c85234c351640efcac62b20487821ba969327e7a#diff-6346c82f50278989f0e85329f1488c325a9bc84afdc883e961bd9ba096384affR80), and that list could instead be a dictionary mapping `index`es on `Container` objects (and the object would not have an `index` attribute). The same would then apply to the `container_type` argument.

In other words, what you need is something like
```python
container_list = {
    index: Container(container["image"])
    for index, container in enumerate(spec["containers"])
]
init_container_list = [
    index: Container(container["image"])
    for index, container in enumerate(spec.get("initContainers", []))
]
```
and then merge them into something with a combined look-up key `(container_type, index)`. Or directly:
```python
container_list = {
    (container_type, index): Container(container["image"])
    for container_type in ['containers', 'initContainers']
    for index, container in enumerate(spec.get(container_type, []))
]
```

You can then do your comparison with `if container in ...parent_containers.values()`.

Once you have this structure, I'm not sure whether the `Container` class is still needed.

Reply:
i implemented @peterthomassen's suggestion and removed the `Container` wrapper.